### ISSUE TITLE
Implement observer chain

### ISSUE THREAD

**Issue by themue** [status-im/go-ethereum#46]
Implement `ObserverChain` for general purpose event logging. Its simplest use case is collecting diagnostic/statistical data. In addition to diagnostic purposes some of this data may also be post-processed with chain filters to provide service quality/capacity/availability estimates for clients.

An observer chain is basically a *personal* blockchain that is generated by individual nodes and validated by a single signature. Each block contains an Ethereum trie structure that is used as a general purpose key/value store. Interpretation of its contents (called *statements*) is application-specific, there are no general rules applying to its contents. Application-specific state transition rules can be defined inside the trie itself (we call these voluntary rules *promises*).

Whether the statement tries of subsequent blocks are interpreted as independent key/value stores (each new statement recorded in one block only) or an *evolving* database where trie entries are inherited from the previous statement trie is up to the specific application. Separate key ranges can be handled differently too. It is also allowed to leave data in the trie forever and just forget entire subtries that are no longer relevant to anyone. Ensuring data availability is the responsibility of everyone who might want to prove the existence of a certain statement later.

```
// NewObserverChain inits structure, reads existing chain head from db if possible.
func NewObserverChain(db ethdb.Database) *ObserverChain

func (o *ObserverChain) GetHead() *ObserverBlock

func (o *ObserverChain) GetBlock(index uint64) *ObserverBlock

// LockAndGetTrie locks trie mutex and gets r/w access to the current observer trie.
func (o *ObserverChain) LockAndGetTrie() *trie.Trie

// UnlockTrie unlocks trie mutex.
func (o *ObserverChain) UnlockTrie()

// CreateBlock commits current trie and seals a new block; continues using the same
// trie (values are persistent, we will care about garbage collection later).
func (o *ObserverChain) CreateBlock() *ObserverBlock

// AutoCreateBlocks creates a new block periodically until chain is closed; non-blocking, 
// starts a goroutine.
func (o *ObserverChain) AutoCreateBlocks(period time.Duration)

func (o *ObserverChain) Close()
```

The observer blocks have the following fields:

```
PrevHash	common.Hash
Number	        uint64
UnixTime	uint64
TrieRoot	common.Hash // root hash of a trie.Trie structure that is updated for every new block
SignatureType	string	// "ECDSA"
Signature	[]byte	// 65-byte ECDSA signature
```

Note: Signature is based on the hash of the RLP encoding of the struct while the `Signature` field is set to nil.

For signing with ECDSA see example here: https://github.com/ethereum/go-ethereum/blob/master/les/protocol.go#L144  

Observer chains have to follow these general rules:

- numbers must be consecutive
- each number must be used only once (rolling back and forking are forbidden)
- timestamps must be monotonic

**Comment by themue** [comment]
Implementation of OC is almost done, open:

- [ ] Cleanup after `Close()`
- [ ] Question: How to handle locked tries on block creation or closing, they are out of control?
- [ ] More unit tests.

**Comment by themue** [comment]
Asked Zsolt for review.

**Comment by themue** [comment]
Pushed changes after review and pinged Zsolt.
